### PR TITLE
perf: replace 'input_to_float8' with 'sgl_per_tensor_quant_fp8' (outdated)

### PR DESCRIPTION
**Pull request by saltyfish66**
## Motivation
I am optimizing the performance of model DeepSeek-R1(671B) on 2x8xH20. 
I used nvidia nsys to collect CUDA kernel runtime timeline during model infer and found that many trivial kernels were executed during the model execution.
After some analysis, I found that these trivial kernels are due to a function `input_to_float8` in MLA.
After searching for a while, I found that a kernel named `sgl_per_tensor_quant_fp8` with the same functionality as `input_to_fp8` has been implemented in sgl_kernel.
I replaced `input_to_fp8` with `sgl_per_tensor_quant_fp8` and got some throughtput improvement.

## Modifications

Replace `input_to_float8` in `python/sglang/srt/layers/quantization/fp8_utils.py` with `sgl_per_tensor_quant_fp8`.

## Accuracy Test
In the accuracy test, I found that the precision of function `input_to_float8` is not as good as `sgl_per_tensor_quant_fp8`.
So I wrote function `input_to_float8_high_precision` as ground truth. Function `input_to_float8_high_precision` converts some bf16 tensors to float32 type for calculation based on function `input_to_float8`.
I only ran the test cases on Nvidia H20, which means `_is_hip` was always = False. But I found that `sgl_per_tensor_quant_fp8` already has code logic for handling when `_is_hip == True`.

```
import torch
from typing import Tuple

_is_hip = False


def input_to_float8(
    x: torch.Tensor, dtype: torch.dtype = torch.float8_e4m3fn
) -> Tuple[torch.Tensor, torch.Tensor]:
    """This function quantizes input values to float8 values with tensor-wise quantization."""
    finfo = torch.finfo(dtype)
    min_val, max_val = x.aminmax()
    amax = torch.maximum(min_val.abs(), max_val.abs()).clamp(min=1e-12)
    fp8_max = finfo.max
    if _is_hip:
        dtype = torch.float8_e4m3fnuz
        fp8_max = 224.0
    scale = fp8_max / amax
    x_scl_sat = (x * scale).clamp(min=-fp8_max, max=fp8_max)
    return x_scl_sat.to(dtype).contiguous(), scale.float().reciprocal()


def input_to_float8_high_precision(
    x: torch.Tensor, dtype: torch.dtype = torch.float8_e4m3fn
) -> Tuple[torch.Tensor, torch.Tensor]:
    """This function quantizes input values to float8 values with tensor-wise quantization."""
    finfo = torch.finfo(dtype)
    min_val, max_val = x.aminmax()
    amax = torch.maximum(min_val.abs(), max_val.abs()).clamp(min=1e-12)
    fp8_max = finfo.max
    if _is_hip:
        dtype = torch.float8_e4m3fnuz
        fp8_max = 224.0
    ##### modify here
    # scale = fp8_max / amax
    scale = fp8_max / amax.float()
    x = x.float()
    #####
    x_scl_sat = (x * scale).clamp(min=-fp8_max, max=fp8_max)
    return x_scl_sat.to(dtype).contiguous(), scale.float().reciprocal()


def input_to_float8_after_mr(
    x: torch.Tensor, dtype: torch.dtype = torch.float8_e4m3fn
) -> Tuple[torch.Tensor, torch.Tensor]:
    """This function quantizes input values to float8 values with tensor-wise quantization."""
    from sgl_kernel import sgl_per_tensor_quant_fp8

    x = x.contiguous()
    output = torch.empty_like(x, device=x.device, dtype=dtype)
    scale = torch.zeros(1, device=x.device, dtype=torch.float32)

    sgl_per_tensor_quant_fp8(x, output, scale, is_static=False)
    return output, scale.squeeze()


input_dtypes = [torch.float16, torch.bfloat16]
shapes = [
    (16, 16, 8, 16),
    (128, 128, 256),
    (23, 543, 764),
    (12, 1024, 64),
    (36, 56),
    (1024,),
]

for input_dtype in input_dtypes:
    for shape in shapes:
        print(f"testing {input_dtype=}, {shape=}")
        x = torch.rand(shape, dtype=input_dtype, device="cuda:0")
        # can't pass accuracy test if use input_to_float8 as groud truth
        # expected_q, expected_s = input_to_float8(x)
        expected_q, expected_s = input_to_float8_high_precision(x)
        actual_q, actual_s = input_to_float8_after_mr(x)

        torch.testing.assert_close(actual_s, expected_s)
        torch.testing.assert_close(actual_q.float(), expected_q.float(), rtol=1e-2, atol=1e-2)

torch.cuda.synchronize()
print("test ok")
```

## Benchmark
run DeepSeek-R1-671B on 2x8xH20
```
Launch Server:
# closed deepgemm
export SGL_ENABLE_JIT_DEEPGEMM=0
python -u -m sglang.launch_server \
--model-path /sgl-workspace/DeepSeek-R1 \
--nnodes 2 --trust-remote-code  \
--served-model-name deepseek-r1 \
--dist-init-addr xxxx:5000 \
--node-rank ${NODE_RANK} \
--host 0.0.0.0 --port 8080 \
--tp 16 \
--schedule-policy fcfs \
--disable-overlap-schedule \
--context-length 65536 \
--chunked-prefill-size 32768 \
--max-prefill-tokens 32768 \
--mem-fraction-static 0.85 \
--enable-flashinfer-mla \
--cuda-graph-bs 1 2 4 8 16 32 64 96 128 --cuda-graph-max-bs 128

Benchmark:
For prefilling:
python3 -m sglang.bench_one_batch_server --model None --skip-warmup \
--base-url http://127.0.0.1:8080 --batch-size ${batch_size} --input-len ${input_len} --output-len 0
For decoding:
python3 -m sglang.bench_one_batch_server --model None --skip-warmup \
--base-url http://127.0.0.1:8080 --batch-size ${batch_size} --input-len 1 --output-len 256
```

Benchmark Result:
### For Prefilling Phase
|name|batchsize| input-len| output-len| total latency(s)| decode throughput(token/s)| (input + output) throughput|
|-----|-----|-----|-----|-----|-----|-----|
|before mr|1|4096|0|0.46|0.00|8982.20|
|after mr|1|4096|0|0.46|0.00|8952.61|
|before mr|1|8192|0|0.86|0.00|9543.61|
|after mr|1|8192|0|0.86|0.00|9578.07|
|before mr|1|16384|0|1.80|0.00|9088.08|
|after mr|1|16384|0|1.80|0.00|9086.09|
|before mr|8|4096|0|2.96|0.00|11087.44|
|after mr|8|4096|0|2.95|0.00|11097.90|
|before mr|8|8192|0|6.25|0.00|10492.83|
|after mr|8|8192|0|6.24|0.00|10496.98|
|before mr|8|16384|0|13.86|0.00|9454.62|
|after mr|8|16384|0|13.89|0.00|9439.45|

Conclusion: Almost no impact during prefill phase.

### For Decoding Phase
|name|batchsize| input-len| output-len| total latency(s)| decode throughput(token/s)| (input + output) throughput|
|-----|-----|-----|-----|-----|-----|-----|
|before mr|1|1|256|9.54|26.83|26.93|
|after mr|1|1|256|9.07|28.23|28.35|
|before mr|16|1|256|12.25|334.34|335.65|
|after mr|16|1|256|11.57|354.11|355.50|
|before mr|32|1|256|15.15|540.57|542.68|
|after mr|32|1|256|13.71|597.65|599.99|
|before mr|64|1|256|17.80|920.41|924.00|
|after mr|64|1|256|15.98|1025.12|1029.12|

Conclusion: 5-10% throughput improvement during the decoding phase.

### nsys timeline
Run DeepSeek-R1-671B on 2x8xH20.
Lauched Server with tp=16 and dp=16, infer with batchsize=256(batchsize=16 on each dp).
Use nsys session to capture CUDA kernels timeline during decoding phase.
Obviously, after MR, `input_to_float8` generates fewer CUDA kernels and shorter latency.

image

Note: After MR. `input_to_float8` generate 4 kernels: torch.contiguous, torch.zeros(fill kernel), per_tensor_absmax_kernel and per_tensor_quant_fp8_kernel.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Review by BBuf — APPROVED** [review]
LGTM! Can you add a gsm8k acc test?

**Comment by saltyfish66** [comment]
> LGTM! Can you add a gsm8k acc test?

OK, I will add acc tests later.
